src/dmd/root/file.d

```diff
@@ -23,82 +23,70 @@ import dmd.root.filename;
 import dmd.root.rmem;
 import dmd.utils;
 
-/***********************************************************
- */
-struct File
+/// Owns a (rmem-managed) file buffer.
+struct FileBuffer
 {
-    int _ref; // != 0 if this is a reference to someone else's buffer
-    ubyte* buffer; // data for our file
-    size_t len; // amount of data in buffer[]
-    const(FileName) name; // name of our file
+    ubyte[] data;
 
-nothrow:
-    extern (D) this(const(char)* n)
-    {
-        this(n.toDString());
-    }
+    this(this) @disable;
 
-    extern (D) this(const(char)[] n)
+    ~this() pure nothrow
     {
-        _ref = 0;
-        buffer = null;
-        len = 0;
-        name = FileName(n);
+        mem.xfree(data.ptr);
     }
 
-    extern (C++) static File* create(const(char)* n)
+    /// Transfers ownership of the buffer to the caller.
+    ubyte[] extractData() pure nothrow @nogc @safe
     {
-        return new File(n.toDString());
+        auto result = data;
+        data = null;
+        return result;
     }
 
-    extern (C++) ~this()
+    extern (C++) static FileBuffer* create()
     {
-        if (buffer)
-        {
-            if (_ref == 0)
-                mem.xfree(buffer);
-            version (Windows)
-            {
-                if (_ref == 2)
-                    UnmapViewOfFile(buffer);
-            }
-        }
+        return new FileBuffer();
     }
+}
 
-    extern (C++) const(char)* toChars() const pure nothrow @safe
+///
+struct File
+{
+    ///
+    static struct ReadResult
     {
-        return name.toChars();
-    }
+        bool success;
+        FileBuffer buffer;
 
-    const(char)[] toString() const nothrow pure @safe
-    {
-        return name.toString();
+        /// Transfers ownership of the buffer to the caller.
+        ubyte[] extractData() pure nothrow @nogc @safe
+        {
+            return buffer.extractData();
+        }
     }
 
-    /**
-     * Read the full content of a file
-     *
-     * Returns:
-     *   `true` if there was an error
-     */
-    extern (C++) bool read()
+nothrow:
+    /// Read the full content of a file.
+    extern (C++) static ReadResult read(const(char)* name)
     {
-        if (len)
-            return false; // already read the file
+        ReadResult result;
 
         import core.stdc.string : strcmp;
-        const(char)* name = this.name.toChars();
+        if (strcmp(name, "__main.d") == 0)
```

Comment:
You're saying gdc itself uses a file named `__main.d`, without any path? My goal here was to move the special files to one place, instead of handling `__stdin.d` here and setting up `__main.d` in `mars.d` (the only place where the file buffer was set manually from outside and then `File.read()` being a no-op).

Reply:
```
$ echo "import std.stdio; int main() { writeln(\"Hello\"); return 42; }" > __main.d
$ ./generated/linux/release/64/dmd __main.d -ofmain
$ ./main 
$ echo $?
0
```
How exactly is this desirable behaviour?

IMO, root/file.d should just be a generic wrapper around a file, it should not have any compiler-specific logic embedded within.

Reply:
This applies to `__stdin.d` as well then.

Reply:
=> #9795